### DISCUSSION TITLE
SQLPage whishlist/roadmap

### DISCUSSION THREAD

**Discussion by f8dca** [sqlpage/SQLPage#87] [Ideas]
## Here is my list of possible enhancements and features that I would love to see in SQLPage

(in no particular order)

- Some sort of an include directive to be able to reuse .sql files in multiple pages.
- TLS support
- Table component with built in CRUD support
- Markdown based tooltips on everything
- A Table component that is expandable / collapsible to deal with wider tables
- Set the working directory for the program as a startup parameter
- A user upload component to upload files ,validate against a schema and insert it into a temp table
- Support to temporary tables on per session basis
- Enhance user authentication component to support basic/ Digest / Oauth / Saml / SSO
- API component to call external APIs with proper security, CORS support and session management.
- Connect to a database based on user provided credentials
- Logging to syslog
- First class support to run SQLPage as a service on Linux/Windows  
(currently works with NSSP but would be nice not to use it)
- Support to read and write Local and Global variables as well as Declare statements.
- URI parameter signing or other protection mechanism to prevent unauthorized passing of parameters directly to .sql pages.  
(HMAC signing  the request ?)
- A basic web based status page showing DB connection, session and other metrics.  
(Can be optionally turned on/off - default is off)
- A directive to allow seeing the content of an .sql file in a read only format  
(Optional. default must be off) 
- Add a section to the website for user contributed / 3rd party .handelbars components.
- Add -v parameter to show the version number of the program
- Better code snippet visuals with copy to clopboard icon

**Comment by lovasoa** [comment]
Wow, that's a great list ! Would you be interested in me guiding you to making your first contributions to sqlpage ?

**Comment by f8dca** [comment]
I'm setting up a cloned repo of SQLPage for myself very soon. Expect a few basic PR from me - mostly focusing on documentation first but maybe later a custom mermaid.js component or sample API calls. I consider myself more of a database guy with zero rust knowledge but I have a lot of system integration and architecture experience that might be helpful for your project.

Also my wishlist is just a subjective list of one person - with priorities based on my own use-case scenarios. I hope that more users will come forward with their own list. Once there is enough data a proper development roadmap can be built by combining the most requested features.

This also leads to another important point: Architecturally SQLPage is probably not in its final form. There are a few missing elements, such as include directives, that can be implemented many ways but needs to be decided to ensure consistency.

**Comment by lovasoa** [comment]
Several of these points have already come up in discussions with other users. The next big things on my list are OIDC authentication and multi database. For the latter, I'm still not sure how it should work...

**Comment by f8dca** [comment]
Multi database is purposely not on my list - I think this is a problem that can be solved leveraging the capabilities of the database server and no need to overcomplicate SQLPage with an inferior workaround.

I already have a working solution to pull PostgreSQL server data via my MSSQL. I'll try to remove the company specifics and turn it into a some sort of documentation for SQLPage.

**Comment by lovasoa** [comment]
> - Connect to a database based on user provided credentials

Ideally, I'd like to find a good design that would allow implementing multi-database and dynamic user selection as a part of the same feature.

**Comment by lovasoa** [comment]
> I already have a working solution to pull PostgreSQL server data via my MSSQL. I'll try to remove the company specifics and turn it into a some sort of documentation for SQLPage.

That would be great ! Any contribution, even small, to https://github.com/lovasoa/SQLpage/tree/main/examples is very welcome.

**Comment by powderdba** [comment]
To tag on to this - is there a way for MSSQL to use windows authentication instead of passing a username/password? So any domain pre-authorized users can just open the pages and it works without using a specific username/password for the app

**Comment by powderdba** [comment]
It might already do that and I just haven't seen how to do it :) I've tried looking though I swear 😁 Also - love SQLPage! perfect for me 😄

**Comment by jgranduel** [comment]
Could the CLI be more classical with options, instead of defaulting to starting the HTTP server?
`sqlpage --help, --version`
So maybe `sqlpage start | check | ...`

Could you display the full urls like does [miniserve](https://github.com/svenstaro/miniserve) (for instance) ?
Useful for shell clickable URL ;-) !
```
> miniserve .
miniserve v0.22.0
Bound to [::]:8080, 0.0.0.0:8080
Serving path /path/to/sqlpage
Available at (non-exhaustive list):
    http://127.0.0.1:8080
    http://<IP>:8080
    http://[::1]:8080

Quit by pressing CTRL-C
```

**Comment by lovasoa** [comment]
About pagination, I answered here: https://github.com/lovasoa/SQLpage/issues/84

**Comment by lovasoa** [comment]
I'm moving this to a discussion, but keeping track of it !

**Comment by lovasoa** [comment]
v0.12 beta is being released, with cross-database support for variables, which opens the door to many interesting new usecases !

[SQLPage custom functions](https://sql.ophir.dev/functions.sql) used to be limited to very particular cases, since they could take only sqlpage GET and POST variables, and not random sql expressions as arguments. Variables change that, and allow using data from the database as function inputs.

```sql
SET myvar = (SELECT x FROM y);
SELECT sqlpage.myfunc($myvar);
```


This unlocks creating new functions such as `sqlpage.fetch` for http requests, or `sqlpage.exec` to shell out to external programs.

**Comment by lovasoa** [comment]
I have started work  on TLS support for Microsoft SQL server connections: https://github.com/lovasoa/sqlx-oldapi/pull/8
